naga/src/compact/functions.rs

```diff
@@ -4,6 +4,8 @@ use super::{FunctionMap, ModuleMap};
 pub struct FunctionTracer<'a> {
     pub function: &'a crate::Function,
     pub constants: &'a crate::Arena<crate::Constant>,
+    pub types: &'a crate::UniqueArena<crate::Type>,
+    pub resolve_context: crate::proc::ResolveContext<'a>,
```

Comment:
I don't think this is the right approach. We talked in person about how this works, and I see that it does accomplish the goal. But as much as possible, I would really like to keep the compactor a dumb structural algorithm: you give it this DAG-gy module thing, and it shakes out all the disconnected pieces without looking too hard, and that's it. Running the typifier in compaction introduces subtleties that are not the direction that I would like to see things going.

Instead, as discussed, let's try having the WGSL front end simply not add abstract-typed constants to the function's named expressions table at all. The hope is that this would make compaction just remove them naturally, since constant evaluation should have left them all unused.

Reply:
The avoid adding them to named expressions trick seemed to work. Patch is updated. It felt trivial enough just to keep it in the same PR